src/base/memory_monitor_sr.hpp

```diff
@@ -0,0 +1,234 @@
+
+    public:
+      mmon_module (const char *module_name, const MMON_COMP_INFO *info);
+      mmon_module () {} /* for dummy */
+      mmon_module (const mmon_module &) = delete;
+      mmon_module (mmon_module &) = delete;
+
+      virtual ~mmon_module () {}
+
+      mmon_module &operator = (const mmon_module &) = delete;
+      mmon_module &operator = (mmon_module &) = delete;
+
+      /* function */
+      virtual int aggregate_stats (const MMON_MODULE_INFO &info);
+      void add_stat (uint64_t size, int comp_idx, int subcomp_idx, bool init, bool expand);
+      void sub_stat (uint64_t size, int comp_idx, int subcomp_idx, bool init);
+      int add_component (const char *name);
+      inline int MAKE_STAT_IDX_MAP (int comp_idx, int subcomp_idx);
+
+    private:
+      std::string m_module_name;
+      MMON_STAT m_stat;
+      std::vector<std::unique_ptr<mmon_component>> m_component;
+      std::vector<int> m_comp_idx_map;
+  };
+
+
+  class memory_monitor
+  {
+    public:
+      memory_monitor (const char *server_name)
+	: m_aggregater (this), m_server_name (server_name) {}
+      memory_monitor (const memory_monitor &) = delete;
+      memory_monitor (memory_monitor &) = delete;
+
+      ~memory_monitor ();
+
+      memory_monitor &operator = (const memory_monitor &) = delete;
+      memory_monitor &operator = (memory_monitor &) = delete;
+
+      /* function */
+      int add_stat (THREAD_ENTRY *thread_p, MMON_STAT_ID stat_id, uint64_t size, bool expand);
+      int sub_stat (THREAD_ENTRY *thread_p, MMON_STAT_ID stat_id, uint64_t size);
+      int resize_stat (THREAD_ENTRY *thread_p, MMON_STAT_ID stat_id, uint64_t old_size, uint64_t new_size);
+      int move_stat (THREAD_ENTRY *thread_p, MMON_STAT_ID src, MMON_STAT_ID dest, uint64_t size);
+      int aggregate_module_info (MMON_MODULE_INFO *info, int module_index);
+      int aggregate_tran_info (MMON_TRAN_INFO *info, int tran_count);
+      inline int COMP_IDX_FROM_COMP_IDX_MAP (int comp_idx_map_val);
+      inline int SUBCOMP_IDX_FROM_COMP_IDX_MAP (int comp_idx_map_val);
+
+    private:
+      class aggregater
+      {
+	public:
+	  aggregater (memory_monitor *mmon);
+	  aggregater (const aggregater &) = delete;
+	  aggregater (aggregater &) = delete;
+
+	  ~aggregater () {}
+
+	  aggregater &operator = (const aggregater &) = delete;
+	  aggregater &operator = (aggregater &) = delete;
+
+	  /* function */
+	  int get_server_info (const MMON_SERVER_INFO &info);
+	  int get_module_info (const MMON_MODULE_INFO &info, int module_index);
+	  int get_transaction_info (const MMON_TRAN_INFO &info, int tran_count);
+
+	private:
+	  /* backlink of memory_monitor class */
+	  memory_monitor *m_mmon;
+      };
+
+    private:
+      std::string m_server_name;
+      std::atomic<uint64_t> m_total_mem_usage;
+      aggregater m_aggregater;
+
+      mmon_module *m_module[MMON_MODULE_LAST] =
+      {
+	new mmon_module ()				  /* dummy for aligning module index (1 ~) */
```

Comment:
I'm against using "dummy". It leads to potential errors.
1. What if someone accesses m_module[0]?
2. For this, we have to open `mmon_module ()`. The behavior of the instance created with this ctor is not defined. Or we have to define it, at least with assert().
3.  We waste memory for this.

I recommend converting the index if the index from the user-side must start with 1. In other words, use `index=index-1` when a client inputs the index.

Reply:
I don't think we will have access to [0] through structural defense, but I agree with your point regarding human error and the occurrence of unnecessary memory allocation. I will proceed as suggested.